### PR TITLE
Fix Mac M1 build

### LINKED ISSUE #16230: [Build] Unable to use CoreML: This binary was not compiled with CoreML support.
### Describe the issue

Hello, I am using the dependency - 

```
    <dependency>
      <groupId>com.microsoft.onnxruntime</groupId>
      <artifactId>onnxruntime</artifactId>
      <version>1.15.0</version>
    </dependency>

```
and with the following code- 
``` java
  var env = OrtEnvironment.getEnvironment();
  var sessionOptions = new OrtSession.SessionOptions();
  sessionOptions.addCoreML();
  var session = env.createSession("model.onnx", sessionOptions);
```

I get error 

```Error code - ORT_INVALID_ARGUMENT - message: This binary was not compiled with CoreML support.```

So I tried building the package myself like this- 
`./build.sh --update --build --build_java --config Release --parallel`

but that gives an error 

```
CMake Error at /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find JNI (missing: JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 AWT)
Call Stack (most recent call first):
  /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindJNI.cmake:569 (find_package_handle_standard_args)
  onnxruntime_java.cmake:11 (find_package)
  CMakeLists.txt:1631 (include)
  ```

`java -version` output- 

```
openjdk version "17.0.7" 2023-04-18
OpenJDK Runtime Environment Homebrew (build 17.0.7+0)
OpenJDK 64-Bit Server VM Homebrew (build 17.0.7+0, mixed mode, sharing)
```

Any suggestion on how to go about fixing it? I would rather use pre-built binary than spend time resolving build issues 

### Urgency

_No response_

### Target platform

macOS

### Build script

`./build.sh --update --build --build_java --config Release --parallel`

### Error / output

`  Could NOT find JNI (missing: JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 AWT)
Call Stack (most recent call first):
  /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /opt/homebrew/Cellar/cmake/3.26.4/share/cmake/Modules/FindJNI.cmake:569 (find_package_handle_standard_args)
  onnxruntime_java.cmake:11 (find_package)
  CMakeLists.txt:1631 (include)`

### Visual Studio Version

_No response_

### GCC / Compiler Version

arm64-apple-darwin22.4.0

### LINKED ISSUE #16496: [Build] Compilation on an Apple M1 Pro produces an illegal instruction
### Describe the issue

Seems that #16082 has introduced a bug when compiling for an Apple M1 machine. The bug stems from a missing inlined-assembly [instruction](https://github.com/microsoft/onnxruntime/commit/b28e927ca4e#diff-2f1ef5fcaaa83277145acadea37a3d9da65e9605664f03efd4ca6208fa191e3bR452). 

Reverting the commit introduces a regression on the test suite.

### Urgency

This is a presumed blocker for a release on an arm64 Apple.

### Target platform

arm64-apple-darwin22.5.0

### Build script

`./build.sh --config Debug --build_shared_lib --parallel --compile_no_warning_as_error --skip_submodule_sync --cmake_extra_defines CMAKE_OSX_ARCHITECTURES=arm64`

### Error / output

```
...

5: 2023-06-27 14:08:02.109 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.109664 [I:onnxruntime:Default, bfc_arena.cc:212 Extend] Allocated memory at 0x128a88000 to 0x128e88000
5: 2023-06-27 14:08:02.110 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.110300 [I:onnxruntime:, session_state_utils.cc:344 SaveInitializedTensors] Done saving initialized tensors
5: 2023-06-27 14:08:02.111 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.111169 [I:onnxruntime:, inference_session.cc:1675 Initialize] Session successfully initialized.
5: 2023-06-27 14:08:02.111 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.111225 [V:onnxruntime:, sequential_executor.cc:534 ExecuteThePlan] Number of streams: 1
5: 2023-06-27 14:08:02.111 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.111319 [V:onnxruntime:, sequential_executor.cc:184 SessionScope] Begin execution
5: 2023-06-27 14:08:02.111 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.111356 [I:onnxruntime:Default, bfc_arena.cc:347 AllocateRawInternal] Extending BFCArena for Cpu. bin_num:13 (requested) num_bytes: 3154176 (actual) rounded_bytes:3154176
5: 2023-06-27 14:08:02.111 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.111374 [I:onnxruntime:Default, bfc_arena.cc:206 Extend] Extended allocation by 8388608 bytes.
5: 2023-06-27 14:08:02.111 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.111385 [I:onnxruntime:Default, bfc_arena.cc:209 Extend] Total allocated bytes: 15728640
5: 2023-06-27 14:08:02.111 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.111395 [I:onnxruntime:Default, bfc_arena.cc:212 Extend] Allocated memory at 0x120000000 to 0x120800000
5: 2023-06-27 14:08:02.112 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.112710 [V:onnxruntime:, sequential_executor.cc:518 ExecuteKernel] stream 0 launch kernel with idx 66
5: 2023-06-27 14:08:02.113 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.113577 [V:onnxruntime:, stream_execution_context.cc:171 RecycleNodeInputs] ort value 53 released
5: 2023-06-27 14:08:02.113 onnxruntime_global_thread_pools_test[44651:1311462] 2023-06-27 14:08:02.113592 [V:onnxruntime:, sequential_executor.cc:518 ExecuteKernel] stream 0 launch kernel with idx 2
5/8 Test #5: onnxruntime_global_thread_pools_test ....***Exception: Illegal  0.27 sec

...

38% tests passed, 5 tests failed out of 8

Total Test time (real) =   5.27 sec

The following tests FAILED:
	 1 - onnxruntime_test_all (ILLEGAL)
	 2 - onnx_test_pytorch_converted (ILLEGAL)
	 3 - onnx_test_pytorch_operator (ILLEGAL)
	 4 - onnxruntime_shared_lib_test (ILLEGAL)
	 5 - onnxruntime_global_thread_pools_test (ILLEGAL)```
```
Running `lldb onnxruntime_mlas_test` provides the following output:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_INSTRUCTION (code=1, subcode=0xd5380608)
    frame #0: 0x00000001001b5458 onnxruntime_mlas_test`MLAS_PLATFORM::MLAS_PLATFORM(this=0x000000010029cd48) at platform.cpp:456:5
   453 	   HasDotProductInstructions = (IsProcessorFeaturePresent(PF_ARM_V82_DP_INSTRUCTIONS_AVAILABLE) != 0);
   454 	#else
   455 	   uint64_t isar0_el1;
-> 456 	   asm("mrs %[reg], ID_AA64ISAR0_EL1\n" : [reg] "=r"(isar0_el1) : :);
   457 	   HasDotProductInstructions = ((isar0_el1 >> 44) & 0xfu) == 0x1u;
   458 	#endif
   459 	
Target 0: (onnxruntime_mlas_test) stopped.
```

Reverting the offending commit introduces the following test cases to fail:
```
[  FAILED  ] 11 tests, listed below:
[  FAILED  ] GraphTransformationTests.FuseConvBnAddMulFloat16
[  FAILED  ] GraphTransformationTests.QuickGelu
[  FAILED  ] GraphTransformationTests.ConstantSharing_ShareFloatOrHalfTypedInitializer
[  FAILED  ] GraphTransformationTests.ConstantSharing_Share2DFloatOrHalfTypedInitializer
[  FAILED  ] GraphTransformationTests.ConstantSharing_ShareFloatAndHalfTypedInitializer
[  FAILED  ] GraphTransformationTests.ConstantSharing_Share2DFloatAndHalfTypedInitializer
[  FAILED  ] Float16_Tests.Mul_16_Test
[  FAILED  ] InverseContribOpTest.two_by_two_float16
[  FAILED  ] CastOpTest.NonStringTypes
[  FAILED  ] CastOpTest.ToString
[  FAILED  ] IsNaNOpTest.IsNaNFloat16
```

### Visual Studio Version

_No response_

### GCC / Compiler Version

clang version 11.1.0

### PR DESCRIPTION
**Pull request by edgchen1**
### Description

1. Add ifndef `__APPLE__` to skip lines which cause EXC_BAD_INSTRUCTION error.

2. Fix floatToHalf/doubleToHalf conversion issue and add tests.
Eigen was updated and the type of `__half_raw.x` changed to `__fp16`. See [this commit]( https://gitlab.com/libeigen/eigen/-/commit/e265f7ed8e59c26e15f2c35162c6b8da1c5d594f). From [here](https://clang.llvm.org/docs/LanguageExtensions.html#half-precision-floating-point), "Operators that expect arithmetic operands immediately promote `__fp16` operands to `float`." Apparently, the `__fp16` result was promoted to `float` before assignment to `uint16_t`. Copying the bytes over with something like `std::bit_cast` does what we want here.

### Motivation and Context

Fix Mac M1 build issues.

Fix #16496.
Fix #16230.

### CODE REVIEW

**Comment by cbourjau** [comment]
It would be great if this could be tested in CI going forward (possibly in a follow-up PR?). In the meantime I'd like to confirm that this patch works for me locally! Thanks!

**Comment by Craigacp** [comment]
This also fixes the SIGILL for me on my 2020 13" M1 MBP. It does expose a test failure in the Java CoreML test as CoreML is less precise than the expected value, I'll put a small patch in for that, though I don't think any of the CI runs have both `--build_java` and `--use_coreml` turned on.

File: onnxruntime/core/mlas/lib/platform.cpp [hunk 1]
```diff
@@ -449,10 +449,13 @@ Return Value:
 
 #if defined(_WIN32)
     HasDotProductInstructions = (IsProcessorFeaturePresent(PF_ARM_V82_DP_INSTRUCTIONS_AVAILABLE) != 0);
-#else
+#elif !defined(__APPLE__)  // The next few lines result in an EXC_BAD_INSTRUCTION runtime error on a M1 Mac so we
+                           // disable it there.
```
Comment (chenfucn) [inline onnxruntime/core/mlas/lib/platform.cpp:453]: The change came from https://github.com/microsoft/onnxruntime/pull/16082

dot product feature is available on almost all apple ARM64 chips, setting this to false for all apple platforms might cause performance problem of the CPU EP on those platforms. The GEMM performance difference would be significant. 
 
Can we try this in apple platforms:
    HasDotProductInstructions = MLAS_CPUIDINFO::GetCPUIDInfo().HasArmNeonDot();

**Review by chenfucn — APPROVED** [review]

File: onnxruntime/core/mlas/lib/platform.cpp (hunk 1 above)
Comment (edgchen1) [reply onnxruntime/core/mlas/lib/platform.cpp:453]: Tried `MLAS_CPUIDINFO::GetCPUIDInfo().HasArmNeonDot()` and did not get any error. Will go with that.

**Review by chenfucn — APPROVED** [review]

**Pull request merged**